File: leetcode/BFS/minimum_genetic_mutation.py
from typing import List
import logging

FORMAT = '%(asctime)s %(funcName)s %(lineno)d %(message)s'
logging.basicConfig(level=logging.DEBUG, format=FORMAT)

class Solution:
    def minMutation(self, start: str, end: str, bank: List[str]) -> int:
        # 广度优先搜索
        bank = set(bank)  # 转换为set, in判断只需O(1)时间
        if end not in bank:  # 目标不可行，直接返回-1
            return -1
        q = [(start, 0)]  # 初始结点以及当前步数
        change = {'A': 'TCG', 'T': 'ACG', 'C': 'ATG', 'G': 'ATC'}  # 每个基因对应的可变换基因
        while q:  # 用队列实现广度优先
            node, step = q.pop(0)
            logging.debug("node: %s, step: %d", node, step)
            if node == end:  # 已经到达目标
                return step
            for i, v in enumerate(node):  # 当前序列的每一个基因
                logging.debug("i %d, v: %s", i, v)
                for j in change[v]:  # 该基因可以改变的方式
                    new = node[:i]+j+node[i+1:]  # 改变后的序列
                    logging.debug("new str: %s", new) 
                    if new in bank:  # 如果该序列可行
                        logging.debug("find: %s", new)
                        q.append((new, step+1))  # 入队，继续广度搜索
                        bank.remove(new)  # 避免重复遍历
        return -1  # 队列空了说明不可达

def main():
    solu = Solution()
    start = "AAAACCCC"
    end = "CCCCCCCC"
    bank = ["AAAACCCA","AAACCCCA","AACCCCCA","AACCCCCC","ACCCCCCC","CCCCCCCC","AAACCCCC","AACCCCCC"]
    ret = solu.minMutation(start, end, bank)
    logging.info("result: %d", ret)

if __name__ == "__main__":
    main()

File: leetcode/BFS/minimum_genetic_mutation.cc
#include <algorithm>
#include <map>
#include <queue>
#include <set>
#include <string>
#include <utility>
#include <vector>

#include "third_party/gflags/include/gflags.h"
#include "third_party/glog/include/logging.h"

using std::map;
using std::set;
using std::vector;
using std::queue;
using std::string;
using std::pair;
using std::find;

class Solution {
 public:
  int minMutation(string start, string end, vector<string>& bank) {
    if (find(bank.begin(), bank.end(), end) == bank.end()) return -1;
    set<string> uniq_bank(bank.begin(), bank.end());
    queue<pair<string, int>> try_step;
    try_step.push(make_pair(start, 0));
    // the char match replace chars
    map<char, string> ralations;
    ralations['A'] = "TCG";
    ralations['T'] = "ACG";
    ralations['C'] = "ATG";
    ralations['G'] = "ATC";
    LOG(INFO) << "end: " << end;
    while (!try_step.empty()) {
      // notice: please use front, not back, because it's bfs.
      pair<string, int> step = try_step.front();
      DLOG(INFO) << "str: " << step.first << ", step: " << step.second;
      try_step.pop();
      if (step.first == end) return step.second;
      for (size_t i = 0; i < step.first.size(); ++i) {
        DLOG(INFO) << "i: " << i << ", v: " << step.first[i];
        for (char item : ralations[step.first[i]]) {
          string new_step = step.first.substr(0, i);
          new_step.append(1, item);
          new_step += step.first.substr(i + 1);
          DLOG(INFO) << "new step: " << new_step;
          auto index = find(uniq_bank.begin(), uniq_bank.end(), new_step);
          if (index != uniq_bank.end()) {
            DLOG(INFO) << "find: " << new_step;
            try_step.push(make_pair(new_step, step.second + 1));
            uniq_bank.erase(index);
          }
        }
      }
    }
    return -1;
  }
};

int main(int argc, char* argv[]) {
  google::InitGoogleLogging(argv[0]);
  gflags::ParseCommandLineFlags(&argc, &argv, false);
  Solution solu;
  string start("AAAACCCC");
  string end("CCCCCCCC");
  vector<string> bank = {"AAAACCCA", "AAACCCCA", "AACCCCCA", "AACCCCCC",
                         "ACCCCCCC", "CCCCCCCC", "AAACCCCC", "AACCCCCC"};
  int ret = solu.minMutation(start, end, bank);
  LOG(INFO) << ret;
  return 0;
}
